packages/dds/merge-tree/src/mergeTree.ts

```diff
@@ -887,69 +900,167 @@ export class MergeTree {
 	 * @param segments - An array of (not necessarily contiguous) segments with increasing ordinals.
 	 */
 	private slideAckedRemovedSegmentReferences(segments: ISegment[]) {
-		// References are slid in groups to preserve their order.
-		let currentSlideDestination: ISegment | undefined;
-		let currentSlideIsForward: boolean | undefined;
+		function anyLocalRef(
+			collection: LocalReferenceCollection,
+			predicate: (pos: LocalReferencePosition) => boolean,
+		): boolean {
+			for (const pos of collection) {
+				if (predicate(pos)) {
+					return true;
+				}
+			}
+
+			return false;
+		}
+
+		function filterLocalRefs(
```

Comment:
i'd make this return an `Iterable<LocalReferencePosition>` and then rewrite using a generator. should be less code and marginally faster.